### PR TITLE
Add some automation helpers

### PR DESCRIPTION
**Pull request by sknopper**
Added:
- DragNDrop Library
- Append pdf to email
- Append text to file
- Download file from link

Please take a look at the changes and get back to me.

LG

### CODE REVIEW

File: AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /// <summary>
+    /// Ranorex User Code collection. A collection is used to publish User Code methods to the User Code library.
+    /// </summary>
+    [UserCodeCollection]
+    public class DragNDropLibrary
+    {
+    	
+    	/// <summary>
+    	/// This is a placeholder text. Please describe the purpose of the
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs]: Missing summary, this is still the original placeholder text. Please provide a description of the method

File: AutomationHelpers/src/UserCodeCollections/FileLibrary.cs [hunk 2]
```diff
@@ -43,6 +43,18 @@ public static void WriteToFile(string text, string filenamePrefix, string fileEx
...
             }
         }
+		
+		// <summary>
+        /// Open an existing file and add a new line of text.
+        /// </summary>
+        /// <param name="text">Text that the file should contain</param>
+        /// <param name="path">Full qualified path to load the file including filename and extension</param>
+        [UserCodeMethod]
+        public static void AppendToExistingFile(string text, string path)
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/FileLibrary.cs]: Rather call "AppendLineToExistingFile" because you explicitly add a NewLine?
Frankly, I would not add the NewLine before the text

File: AutomationHelpers/src/UserCodeCollections/WebLibrary.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    			using (WebClient client = new WebClient()) 
+    			{
+    				try
+    				{
+    					client.DownloadFile(uri, localPath);
+    					string message = String.Format("File successfully downloaded to {0}", new FileInfo(localPath).FullName);
+    					Report.Log(ReportLevel.Success, message);
+    				} catch (Exception e)
+    				{
+    					string message = String.Format("Downloading a file from: {0} failed faile for the following reason: {1}", uri, e.Message);
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/WebLibrary.cs]: Remove duplicate word "faile"

File: AutomationHelpers/src/UserCodeCollections/WebLibrary.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    					Report.Log(ReportLevel.Failure, message);
+    				}
+    				
+    			}
+    		} else {
+                //TODO: what should happen if file exists and we don't want to overwrite it?
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/WebLibrary.cs]: You also just log that the download failed, so this is consistent.
But IMHO throwing and exception or passing on the exception would be more correct, since following code will most probably rely on the downloaded file.
Additionally, the method does not provide a return value, so you don't even know in code that it succeeded or failed, so you cannot react to that -> IMHO go with exception

File: AutomationHelpers/src/UserCodeCollections/EmailLibrary.cs [hunk 5]
```diff
@@ -43,41 +43,74 @@ public static class EmailLibrary
...
+        /// <param name="from">Email sender</param>
+        /// <param name="body">Email message</param>
+        /// <param name="attachment">Path of a file to attach</param>
+        /// <param name="serverHostname">Server hostname</param>
+        /// <param name="serverPort">Server port</param>
+        /// <param name="useSSL">Defines whether SSL is used or not (true or false)</param>
+        /// <param name="username">Username</param>
+        /// <param name="password">Password</param>
+        [UserCodeMethod]
+            public static void SendEmail(
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/EmailLibrary.cs]: Either pass on exceptions or change the return type to bool and let the caller know if sending succeeded or not.
IMHO passing on/throwing exception would be better - although it would be a breaking change now. But rather now than later...

File: AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs (hunk 1 above)
Comment (sknopper) [reply AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs]: Updated, sorry for that.

File: AutomationHelpers/src/UserCodeCollections/EmailLibrary.cs (hunk 5 above)
Comment (sknopper) [reply AutomationHelpers/src/UserCodeCollections/EmailLibrary.cs]: The Email Library now returns a bool, but because the SmtpCLient does return void, it is an assumption. If it fails, there should be an exception which triggers an error log (failed test case).

File: AutomationHelpers/src/UserCodeCollections/FileLibrary.cs (hunk 2 above)
Comment (sknopper) [reply AutomationHelpers/src/UserCodeCollections/FileLibrary.cs]: Renamed.
Because in my case the line break is needed i add a parameter which let the caller decide if the line break should be added.

File: AutomationHelpers/src/UserCodeCollections/WebLibrary.cs (hunk 3 above)
Comment (sknopper) [reply AutomationHelpers/src/UserCodeCollections/WebLibrary.cs]: removed

File: AutomationHelpers/src/UserCodeCollections/WebLibrary.cs (hunk 4 above)
Comment (sknopper) [reply AutomationHelpers/src/UserCodeCollections/WebLibrary.cs]: Its not that easy to go with an exception. The execution should only get into this branch if the file exists (was already downloaded before) and the test creator has decided to not download this file again. In this case the test should also be positive because the file exists and is ready for use.

Please feel free to correct me if i am wrong.

**Comment by sknopper** [comment]
Changes done, please have a look at my comments to the change requests

File: AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs [hunk 6]
```diff
@@ -20,13 +20,13 @@ namespace CheckAutomationHelperIntegration.UserCodeCollections
...
-    	/// <summary>
-    	/// This is a placeholder text. Please describe the purpose of the
-    	/// user code method here. The method is published to the User Code library
-    	/// within a User Code collection.
-    	/// </summary>
-    	[UserCodeMethod]
+
+        /// <summary>
+        /// Picks the source element, drags it to the target element and drop it there.
+        /// <param name="source">The element which should be draged.</param>
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs]: Typo "dragged"

File: AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs [hunk 7]
```diff
@@ -20,13 +20,13 @@ namespace CheckAutomationHelperIntegration.UserCodeCollections
...
-    	
-    	/// <summary>
-    	/// This is a placeholder text. Please describe the purpose of the
-    	/// user code method here. The method is published to the User Code library
-    	/// within a User Code collection.
-    	/// </summary>
-    	[UserCodeMethod]
+
+        /// <summary>
+        /// Picks the source element, drags it to the target element and drop it there.
```
Comment (ahoisl) [inline AutomationHelpers/src/UserCodeCollections/DragNDropLibrary.cs]: Typo "drops"

**Comment by Didddy** [comment]
@sknopper pls remove AutomationHelpers/nuget.exe, guess this was added by mistake

File: AutomationHelpers/packages.config [hunk 8]
```diff
@@ -1,4 +1,4 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="Ranorex.ReportToPDF" version="1.0.9" targetFramework="net40" />
-</packages>
+</packages>
```
Comment (Didddy) [inline AutomationHelpers/packages.config]: We have no clear line-ending-strategy atm :) so pls revert this change

File: AutomationHelpers/RanorexAutomationHelpers.csproj [hunk 9]
```diff
@@ -120,10 +123,13 @@
...
       <id>8331baf6-8378-4cde-b35a-b1ed70e726b7</id>
     </Compile>
+    <Compile Include="src\UserCodeCollections\WebLibrary.cs">
+      <id>ea5ff409-14a0-48a0-9377-50a0778534ba</id>
+    </Compile>
   </ItemGroup>
   <ItemGroup>
     <None Include="packages.config">
-      <id>bd96e9e9-2c23-443c-ac89-905ed797822b</id>
+      <id>3bb61b55-d05d-4220-a24c-74dca35bbd05</id>
```
Comment (Didddy) [inline AutomationHelpers/RanorexAutomationHelpers.csproj]: This seems to be unrelated, pls revert this line change

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 10]
```diff
@@ -151,13 +141,64 @@ private void OnTestSuiteCompletedSendResult(object sender, EventArgs e)
...
+
+                Report.Zip(TestReport.ReportEnvironment, zippedReportFileDirectory, name);
+                Report.Info("Zipped report has been generated.");
+                return string.Format("{0}\\{1}.rxzlog", zippedReportFileDirectory, TestReport.ReportEnvironment.ReportName);
+            }
+
+            return string.Empty;
+        }
+        
+        private void finishTestReport()
```
Comment (Didddy) [inline AutomationHelpers/src/Modules/EmailModule.cs]: Method names have to be uppercase

File: AutomationHelpers/src/Modules/EmailModule.cs [hunk 11]
```diff
@@ -151,13 +141,64 @@ private void OnTestSuiteCompletedSendResult(object sender, EventArgs e)
...
+                pdfModule.PdfName = TestReport.ReportEnvironment.ReportName + ".pdf";
+                pdfModule.Xml = "";
+                pdfModule.Details = "all";
+                return pdfModule.createPDF();
+            }
+
+            return string.Empty;
+        }
+
+        private string[] createReports()
```
Comment (Didddy) [inline AutomationHelpers/src/Modules/EmailModule.cs]: Method names have to be uppercase

File: AutomationHelpers/src/Modules/ReportToPDFModule.cs [hunk 12]
```diff
@@ -59,67 +59,90 @@ void ITestModule.Run()
...
-                    UpdateError();
+                    createPDF();
                 };
 
                 this.registered = true;
             }
         }
 
-        private string ConvertReportToPDF(string pdfName, string xml, string details)
+        public string createPDF()
```
Comment (Didddy) [inline AutomationHelpers/src/Modules/ReportToPDFModule.cs]: Method names have to be uppercase

File: AutomationHelpers/src/Modules/ReportToPDFModule.cs [hunk 13]
```diff
@@ -59,67 +59,90 @@ void ITestModule.Run()
...
+            
             return PDFReportFilePath;
         }
 
         private static string AddPdfExtension(string pdfName)
         {
             return pdfName.EndsWith(".pdf") ? pdfName : pdfName + ".pdf";
         }
+        
+        private void finishReport() {
```
Comment (Didddy) [inline AutomationHelpers/src/Modules/ReportToPDFModule.cs]: Method names have to be uppercase

**Comment by Didddy** [comment]
@sknopper good job and big THX for your contribution!!
